apiservers/portlayer/restapi/handlers/containers_handlers.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handlers
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+
+	"github.com/docker/docker/pkg/namesgenerator"
+	middleware "github.com/go-swagger/go-swagger/httpkit/middleware"
+	"golang.org/x/net/context"
+
+	log "github.com/Sirupsen/logrus"
+
+	"net/http"
+
+	"github.com/vmware/vic/apiservers/portlayer/models"
+	"github.com/vmware/vic/apiservers/portlayer/restapi/operations"
+	"github.com/vmware/vic/apiservers/portlayer/restapi/operations/containers"
+	"github.com/vmware/vic/apiservers/portlayer/restapi/options"
+	"github.com/vmware/vic/metadata"
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/portlayer/exec"
+)
+
+// ContainersHandlersImpl is the receiver for all of the exec handler methods
+type ContainersHandlersImpl struct {
+	handlerCtx *HandlerContext
+}
+
+const (
+	serialOverLANPort = 2377
+)
+
+// Configure assigns functions to all the exec api handlers
+func (handler *ContainersHandlersImpl) Configure(api *operations.PortLayerAPI, handlerCtx *HandlerContext) {
+	api.ContainersCreateHandler = containers.CreateHandlerFunc(handler.CreateHandler)
+	api.ContainersStateChangeHandler = containers.StateChangeHandlerFunc(handler.StateChangeHandler)
+	api.ContainersGetHandler = containers.GetHandlerFunc(handler.GetHandler)
+	api.ContainersCommitHandler = containers.CommitHandlerFunc(handler.CommitHandler)
+
+	handler.handlerCtx = handlerCtx
+}
+
+// CreateHandler creates a new container
+func (handler *ContainersHandlersImpl) CreateHandler(params containers.CreateParams) middleware.Responder {
+	defer trace.End(trace.Begin("Containers.CreateHandler"))
+
+	var err error
+	var name string
+	session := handler.handlerCtx.Session
+
+	ctx := context.Background()
+
+	log.Debugf("Path: %#v", params.CreateConfig.Path)
+	log.Debugf("Args: %#v", params.CreateConfig.Args)
+	log.Debugf("Env: %#v", params.CreateConfig.Env)
+	log.Debugf("WorkingDir: %#v", params.CreateConfig.WorkingDir)
+
+	id := exec.GenerateID().String()
+	// Autogenerate a name if client doesn't specify one
+	if params.Name == nil {
+		name = namesgenerator.GetRandomName(0)
+	} else {
+		name = *params.Name
+	}
+
+	// Init key for tether
+	privateKey, err := rsa.GenerateKey(rand.Reader, 512)
```

Comment:
to some extent it depends on what you're trying to do. At the moment there's no validation of the key - it's present solely because the crypto/ssh code requires a host key to function at all.

We're using ssh primarily for the muxing, out-of-band control channels, and potential for scp rather than for the encryption. Getting the key management/distribution correct requires significantly more thread modelling and thought about the trust domains - for now we assume that the ESX infrastructure is mostly trustworthy.